builtin/providers/azurerm/resource_arm_servicebus_queue.go

```diff
@@ -0,0 +1,229 @@
+	}
+
+	enableBatchedOps := d.Get("enable_batched_operations").(bool)
+	enableExpress := d.Get("enable_express").(bool)
+	enablePartitioning := d.Get("enable_partitioning").(bool)
+	maxSize := int64(d.Get("max_size_in_megabytes").(int))
+	requiresDuplicateDetection := d.Get("requires_duplicate_detection").(bool)
+	supportOrdering := d.Get("support_ordering").(bool)
+
+	parameters.QueueProperties.EnableBatchedOperations = &enableBatchedOps
+	parameters.QueueProperties.EnableExpress = &enableExpress
+	parameters.QueueProperties.EnablePartitioning = &enablePartitioning
+	parameters.QueueProperties.MaxSizeInMegabytes = &maxSize
+	parameters.QueueProperties.RequiresDuplicateDetection = &requiresDuplicateDetection
+	parameters.QueueProperties.SupportOrdering = &supportOrdering
+
+	_, err := client.CreateOrUpdate(resGroup, namespaceName, name, parameters)
+	if err != nil {
+		return err
+	}
+
+	read, err := client.Get(resGroup, namespaceName, name)
+	if err != nil {
+		return err
+	}
+	if read.ID == nil {
+		return fmt.Errorf("Cannot read ServiceBus Queue %s (resource group %s) ID", name, resGroup)
+	}
+
+	d.SetId(*read.ID)
+
+	return resourceArmServiceBusQueueRead(d, meta)
+}
+
+func resourceArmServiceBusQueueRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).serviceBusQueuesClient
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resGroup := id.ResourceGroup
+	namespaceName := id.Path["namespaces"]
+	name := id.Path["queues"]
+
+	resp, err := client.Get(resGroup, namespaceName, name)
+	if err != nil {
+		return fmt.Errorf("Error making Read request on Azure ServiceBus Queue %s: %s", name, err)
+	}
+	if resp.StatusCode == http.StatusNotFound {
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("name", resp.Name)
+	d.Set("resource_group_name", resGroup)
+	d.Set("namespace_name", namespaceName)
+	d.Set("location", azureRMNormalizeLocation(*resp.Location))
+
+	props := resp.QueueProperties
+	d.Set("auto_delete_on_idle", props.AutoDeleteOnIdle)
+	d.Set("default_message_ttl", props.DefaultMessageTimeToLive)
+
+	if props.DuplicateDetectionHistoryTimeWindow != nil && *props.DuplicateDetectionHistoryTimeWindow != "" {
+		d.Set("duplicate_detection_history_time_window", props.DuplicateDetectionHistoryTimeWindow)
+	}
+
+	d.Set("enable_batched_operations", props.EnableBatchedOperations)
+	d.Set("enable_express", props.EnableExpress)
+	d.Set("enable_partitioning", props.EnablePartitioning)
+	d.Set("requires_duplicate_detection", props.RequiresDuplicateDetection)
+	d.Set("support_ordering", props.SupportOrdering)
+
+	maxSize := int(*props.MaxSizeInMegabytes)
+
+	// if the queue is in a premium namespace and partitioning is enabled then the
+	// max size returned by the API will be 16 times greater than the value set
+	if *props.EnablePartitioning {
+		namespace, err := meta.(*ArmClient).serviceBusNamespacesClient.Get(resGroup, namespaceName)
```

Comment:
It needs to retrieve the parent Namespace resource where the queue is running to get the size (Basic, Standard or Premium) to be able to re-calculate the max size property differently for Basic and Standard sizes than for Premium size, when getting the queue info from Azure.

In Azure Service Bus, if partitioning is enabled and maximum queue size is set to N megabytes, when getting the max size property of an existing queue we get N MB for Premium and N*16 MB for Basic and Standard. That is because Basic and Standard are running on shared resources and it creates 16 copies/partitions, which means that total max size is 16 times the value that was set. However, Premium runs on dedicated resources and partitioning is always enabled with 2 partitions with max size split between the them. This means that each partition has max size of N/2 MB and the total max size remains N MB.

So to avoid a problem of discrepancy between the value of the property we set when creating the queue and the value of that property we get when we retrieve it from Azure after it was created, it needs to divide the property value by 16 for queues running in Basic and Standard namespaces.